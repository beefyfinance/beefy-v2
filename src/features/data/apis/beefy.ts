import axios, { AxiosInstance } from 'axios';
import BigNumber from 'bignumber.js';
import { isString } from 'lodash';
import { ChainEntity } from '../entities/chain';
import { TokenEntity } from '../entities/token';
import { VaultEntity } from '../entities/vault';
import { mapValuesDeep } from '../utils/array-utils';
import { featureFlag_simulateBeefyApiError } from '../utils/feature-flags';

interface ApyGovVault {
  vaultApr: number;
}
interface ApyMaxiVault {
  totalApy: number;
}
interface ApyStandard {
  beefyPerformanceFee: number;
  vaultApr: number;
  compoundingsPerYear: number;
  vaultApy: number;
  tradingApr?: number;
  totalApy: number;
  // todo: does it make sense to have fees and apy in the same entities?
  lpFee: number;
}
export type ApyData = ApyGovVault | ApyMaxiVault | ApyStandard;
export function isStandardVaultApy(apy: ApyData): apy is ApyStandard {
  return 'compoundingsPerYear' in apy;
}
export function isGovVaultApy(apy: ApyData): apy is ApyGovVault {
  return 'vaultApr' in apy && !('compoundingsPerYear' in apy);
}
export function isMaxiVaultApy(apy: ApyData): apy is ApyMaxiVault {
  return 'totalApy' in apy && !('compoundingsPerYear' in apy);
}

export interface BeefyAPITokenPricesResponse {
  [tokenId: TokenEntity['id']]: number;
}
export interface BeefyAPIBreakdownResponse {
  [vaultId: VaultEntity['id']]: ApyData;
}

export interface BeefyAPIBuybackResponse {
  // those are of type string but they represent numbers
  [chainId: ChainEntity['id']]: { buybackTokenAmount: BigNumber; buybackUsdAmount: BigNumber };
}

// note that there is more infos but we don't need it
type BeefyAPIVaultsResponse = {
  id: string;
  lastHarvest?: number | string;
}[];

export class BeefyAPI {
  public api: AxiosInstance;
  public data: AxiosInstance;

  constructor() {
    // this could be mocked by passing mock axios to the constructor
    this.api = axios.create({
      baseURL: 'https://api.beefy.finance',
      timeout: 30 * 1000,
    });
    this.data = axios.create({
      baseURL: 'https://data.beefy.finance',
      timeout: 30 * 1000,
    });
  }

  // here we can nicely type the responses
  public async getPrices(): Promise<BeefyAPITokenPricesResponse> {
    if (featureFlag_simulateBeefyApiError('prices')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/prices', { params: { _: this.getCacheBuster('hour') } });
    return res.data;
  }

  // i'm not 100% certain about the return type
  // are those token ids ?
  public async getLPs(): Promise<BeefyAPITokenPricesResponse> {
    if (featureFlag_simulateBeefyApiError('lps')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/lps', { params: { _: this.getCacheBuster('hour') } });
    return res.data;
  }

  public async getBreakdown(): Promise<BeefyAPIBreakdownResponse> {
    if (featureFlag_simulateBeefyApiError('apy')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/apy/breakdown', {
      params: { _: this.getCacheBuster('day') },
    });

    // somehow, all vaultApr are currently strings, we need to fix that before sending
    // the data to be processed
    const data = mapValuesDeep(res.data, (val, key) => {
      if (key === 'vaultApr' && isString(val)) {
        val = parseFloat(val);
      }
      return val;
    });

    return data;
  }

  public async getBuyBack(): Promise<BeefyAPIBuybackResponse> {
    if (featureFlag_simulateBeefyApiError('buyback')) {
      throw new Error('Simulated beefy api error');
    }

    type ResponseType = {
      data: {
        [chainId: ChainEntity['id']]: {
          buybackTokenAmount: string;
          buybackUsdAmount: string;
        };
      };
    };
    const strRes = await this.api.get<ResponseType>('/bifibuyback', {
      params: { _: this.getCacheBuster('hour') },
    });
    // format result with big number as we get strings from the api
    const res: BeefyAPIBuybackResponse = {};
    for (const chainId in strRes.data) {
      res[chainId] = {
        buybackTokenAmount: new BigNumber(strRes.data[chainId].buybackTokenAmount),
        buybackUsdAmount: new BigNumber(strRes.data[chainId].buybackUsdAmount),
      };
    }
    return res;
  }

  /**
   * For now we fetch lastHarvest from the api
   * TODO: fetch this from the contract directly
   */
  public async getVaultLastHarvest(vaultId: VaultEntity['id']): Promise<null | Date> {
    const res = await this.api.get<BeefyAPIVaultsResponse>('/vaults', {
      params: { _: this.getCacheBuster('hour') },
    });

    // const vault = data.filter(vault => vault.id.includes(vaultId));
    const vaultConfig = res.data.find(vault => {
      return vault.id === vaultId;
    });

    if (!vaultConfig) {
      // vault is not harvestable (most probably a gov vault)
      return null;
    }
    if (!('lastHarvest' in vaultConfig)) {
      return null;
    }
    let lh = 0;
    if (isString(vaultConfig.lastHarvest)) {
      lh = parseInt(vaultConfig.lastHarvest);
    } else {
      lh = vaultConfig.lastHarvest;
    }
    if (lh === 0) {
      return null;
    }

    return new Date(lh * 1000);
  }

  // maybe have a local cache instead of this cache busting
  // have to check if this returns browser cache before doing so
  protected getCacheBuster(mode: 'hour' | 'day' = 'day'): number {
    if (mode === 'hour') {
      return Math.trunc(Date.now() / (1000 * 60));
    } else {
      const cache = new Date();
      cache.setMinutes(0, 0, 0);
      return cache.getTime();
    }
  }
}
