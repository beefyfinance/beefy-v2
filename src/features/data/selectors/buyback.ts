import { createSelector } from '@reduxjs/toolkit';
import { BIG_ZERO } from '../../../helpers/format';
import { BeefyState } from '../../../redux-types';

export const selectTotalBuybackUsdAmount = createSelector(
  (state: BeefyState) => state.biz.buyback.byChainId,
  byChainId => {
    let total = BIG_ZERO;
    for (const chainId in byChainId) {
      total = total.plus(byChainId[chainId].buybackUsdAmount);
    }
    return total;
  }
);

export const selectTotalBuybackTokenAmount = createSelector(
  (state: BeefyState) => state.biz.buyback.byChainId,
  byChainId => {
    let total = BIG_ZERO;
    for (const chainId in byChainId) {
      total = total.plus(byChainId[chainId].buybackTokenAmount);
    }
    return total;
  }
);
