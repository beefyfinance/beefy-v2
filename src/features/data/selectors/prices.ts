import { createGlobalDataSelector, hasLoaderFulfilledOnce } from './data-loader-helpers.ts';

export const selectIsPricesAvailable = createGlobalDataSelector('prices', hasLoaderFulfilledOnce);
