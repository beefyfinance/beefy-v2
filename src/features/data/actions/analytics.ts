import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import { getAnalyticsApi } from '../apis/instances';
import type { VaultTimelineAnalyticsEntity } from '../entities/analytics';
import BigNumber from 'bignumber.js';
import type { AnalyticsPriceResponse, TimeBucketType } from '../apis/analytics/analytics-types';
import type { VaultEntity } from '../entities/vault';
import { isFiniteNumber } from '../../../helpers/number';

export interface fetchWalletTimelineFulfilled {
  timeline: VaultTimelineAnalyticsEntity[];
  walletAddress: string;
  state: BeefyState;
}

export const fetchWalletTimeline = createAsyncThunk<
  fetchWalletTimelineFulfilled,
  { address: string },
  { state: BeefyState }
>('analytics/fetchWalletTimeline', async ({ address }, { getState }) => {
  const api = await getAnalyticsApi();

  const userTimeline = await api.getWalletTimeline(address);

  const timeline = userTimeline.map(row => {
    return {
      chain: row.chain,
      datetime: new Date(row.datetime),
      displayName: row.display_name,
      isEol: row.is_eol,
      productKey: row.product_key,
      shareBalance: new BigNumber(row.share_balance),
      shareDiff: new BigNumber(row.share_diff),
      shareToUnderlyingPrice: new BigNumber(row.share_to_underlying_price),
      underlyingBalance: new BigNumber(row.underlying_balance),
      underlyingDiff: new BigNumber(row.underlying_diff),
      underlyingToUsdPrice: isFiniteNumber(row.underlying_to_usd_price)
        ? new BigNumber(row.underlying_to_usd_price)
        : null,
      usdBalance: isFiniteNumber(row.usd_balance) ? new BigNumber(row.usd_balance) : null,
      usdDiff: isFiniteNumber(row.usd_diff) ? new BigNumber(row.usd_diff) : null,
    };
  });

  return { timeline, walletAddress: address.toLowerCase(), state: getState() };
});

interface DataMartPricesFulfilled {
  data: AnalyticsPriceResponse;
  vaultId: VaultEntity['id'];
  timebucket: TimeBucketType;
  walletAddress: string;
  state: BeefyState;
}

interface DataMartPricesProps {
  productKey: string;
  timebucket: TimeBucketType;
  walletAddress: string;
  vaultId: VaultEntity['id'];
}

export const fetchShareToUnderlying = createAsyncThunk<
  DataMartPricesFulfilled,
  DataMartPricesProps,
  { state: BeefyState }
>(
  'analytics/fetchShareToUnderlying',
  async ({ productKey, walletAddress, timebucket, vaultId }, { getState }) => {
    const api = await getAnalyticsApi();
    const data = await api.getVaultPrices(productKey, 'share_to_underlying', timebucket);
    return {
      data,
      vaultId,
      timebucket,
      walletAddress: walletAddress.toLocaleLowerCase(),
      state: getState(),
    };
  }
);

export const fetchUnderlyingToUsd = createAsyncThunk<
  DataMartPricesFulfilled,
  DataMartPricesProps,
  { state: BeefyState }
>(
  'analytics/fetchUnderlyingToUsd',
  async ({ productKey, timebucket, walletAddress, vaultId }, { getState }) => {
    const api = await getAnalyticsApi();
    const data = await api.getVaultPrices(productKey, 'underlying_to_usd', timebucket);
    return {
      data,
      vaultId,
      timebucket,
      walletAddress: walletAddress.toLocaleLowerCase(),
      state: getState(),
    };
  }
);
