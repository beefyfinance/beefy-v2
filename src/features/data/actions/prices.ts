import { createAsyncThunk } from '@reduxjs/toolkit';
import { BeefyAPIBuybackResponse, BeefyAPITokenPricesResponse } from '../apis/beefy';
import { getBeefyApi } from '../apis/instances';

export const fetchAllPricesAction = createAsyncThunk<BeefyAPITokenPricesResponse, {}>(
  'prices/fetchAllPricesAction',
  async () => {
    const api = getBeefyApi();
    const [prices, lpPrices] = await Promise.all([api.getPrices(), api.getLPs()]);
    return {
      ...prices,
      ...lpPrices,
    };
  }
);

export const fetchBeefyBuybackAction = createAsyncThunk<BeefyAPIBuybackResponse, {}>(
  'prices/fetchBeefyBuybackAction',
  async () => {
    const api = getBeefyApi();
    return api.getBuyBack();
  }
);
