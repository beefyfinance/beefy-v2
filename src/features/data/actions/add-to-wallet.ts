import type { ChainEntity } from '../entities/chain';
import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { TokenEntity } from '../entities/token';
import { selectTokenByAddress } from '../selectors/tokens';
import { getSingleAssetSrc } from '../../../helpers/singleAssetSrc';

export type AddToWalletParams = {
  chainId: ChainEntity['id'];
  tokenAddress: TokenEntity['address'];
  customIconUrl?: string;
};

export type AddTokenToWalletPayload = {
  token: TokenEntity;
  iconUrl: string;
};

function getTokenIconUrl(
  id: TokenEntity['id'],
  chainId: TokenEntity['chainId'],
  customUrl?: string
): string {
  if (customUrl) {
    return customUrl;
  }

  const assetSrc = getSingleAssetSrc(id, chainId);
  if (assetSrc) {
    return `${window.location.origin}${assetSrc}`;
  }

  return '';
}

export const addTokenToWalletAction = createAsyncThunk<
  AddTokenToWalletPayload,
  AddToWalletParams,
  { state: BeefyState }
>('addToWallet/open', async ({ chainId, tokenAddress, customIconUrl }, { getState }) => {
  const state = getState();
  const token = selectTokenByAddress(state, chainId, tokenAddress);
  const iconUrl = getTokenIconUrl(token.id, token.chainId, customIconUrl);

  return {
    token,
    iconUrl,
  };
});
