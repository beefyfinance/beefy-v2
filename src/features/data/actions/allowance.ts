import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { FetchAllAllowanceResult } from '../apis/allowance/allowance-types';
import { getAllowanceApi } from '../apis/instances';
import type { ChainEntity } from '../entities/chain';
import type { TokenErc20 } from '../entities/token';
import type { VaultGov, VaultStandard } from '../entities/vault';
import { isGovVault } from '../entities/vault';
import { selectBoostById, selectBoostsByChainId } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectVaultByChainId, selectVaultById } from '../selectors/vaults';
import { selectWalletAddress } from '../selectors/wallet';

interface ActionParams {
  chainId: ChainEntity['id'];
  walletAddress: string;
}

export interface FetchAllAllowanceFulfilledPayload {
  chainId: ChainEntity['id'];
  data: FetchAllAllowanceResult;
}

export const fetchAllAllowanceAction = createAsyncThunk<
  FetchAllAllowanceFulfilledPayload,
  ActionParams,
  { state: BeefyState }
>('allowance/fetchAllAllowanceAction', async ({ chainId, walletAddress }, { getState }) => {
  const state = getState();

  const userAddress = walletAddress || selectWalletAddress(state);
  const chain = selectChainById(state, chainId);
  const api = await getAllowanceApi(chain);

  // maybe have a way to retrieve those easily
  const boosts = selectBoostsByChainId(state, chainId).map(vaultId =>
    selectBoostById(state, vaultId)
  );
  const allVaults = selectVaultByChainId(state, chainId).map(vaultId =>
    selectVaultById(state, vaultId)
  );
  const standardVaults: VaultStandard[] = [];
  const govVaults: VaultGov[] = [];
  for (const vault of allVaults) {
    if (isGovVault(vault)) {
      govVaults.push(vault);
    } else {
      standardVaults.push(vault);
    }
  }

  // always re-fetch state as late as possible
  const data = await api.fetchAllAllowances(
    // always re-fetch state as late as possible
    getState(),
    standardVaults,
    govVaults,
    boosts,
    userAddress
  );
  return { chainId, data };
});

interface FetchAllowanceActionParams {
  chainId: ChainEntity['id'];
  spenderAddress: string;
  tokens: TokenErc20[];
  walletAddress: string;
}

export const fetchAllowanceAction = createAsyncThunk<
  FetchAllAllowanceFulfilledPayload,
  FetchAllowanceActionParams,
  { state: BeefyState }
>(
  'allowance/fetchAllowanceAction',
  async ({ chainId, spenderAddress, tokens, walletAddress }, { getState }) => {
    const state = getState();
    const userAddress = walletAddress || selectWalletAddress(state);
    const chain = selectChainById(state, chainId);
    const api = await getAllowanceApi(chain);

    const allowanceRes =
      userAddress && spenderAddress
        ? await api.fetchTokensAllowance(getState(), tokens, userAddress, spenderAddress)
        : [];

    return { chainId, data: allowanceRes };
  }
);
