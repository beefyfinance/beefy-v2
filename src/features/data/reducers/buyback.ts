import { createSlice } from '@reduxjs/toolkit';
import BigNumber from 'bignumber.js';
import { fetchBeefyBuybackAction } from '../actions/prices';
import { ChainEntity } from '../entities/chain';

/**
 * State containing APY infos indexed by vault id
 */
export interface BuybackState {
  byChainId: {
    [byChainId: ChainEntity['id']]: { buybackTokenAmount: BigNumber; buybackUsdAmount: BigNumber };
  };
}
export const initialBuybackState: BuybackState = {
  byChainId: {},
};

export const buybackSlice = createSlice({
  name: 'buyback',
  initialState: initialBuybackState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(fetchBeefyBuybackAction.fulfilled, (sliceState, action) => {
      // states have compatible types for now
      sliceState.byChainId = action.payload;
    });
  },
});
