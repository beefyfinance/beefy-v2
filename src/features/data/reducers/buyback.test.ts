import BigNumber from 'bignumber.js';
import { BeefyState } from '../../../redux-types';
import { fetchBeefyBuybackAction } from '../actions/prices';
import { BeefyAPIBuybackResponse } from '../apis/beefy';
import { selectTotalBuybackTokenAmount, selectTotalBuybackUsdAmount } from '../selectors/buyback';
import { buybackSlice, initialBuybackState } from './buyback';

describe('Buyback slice tests', () => {
  it('should update state on fulfilled buyback action', async () => {
    const payload: BeefyAPIBuybackResponse = {
      bsc: { buybackTokenAmount: new BigNumber('0'), buybackUsdAmount: new BigNumber('0') },
      polygon: {
        buybackTokenAmount: new BigNumber('0.535709913009709997'),
        buybackUsdAmount: new BigNumber('1197.3521608619992880891961859928494'),
      },
      cronos: {
        buybackTokenAmount: new BigNumber('0.503019530037920513'),
        buybackUsdAmount: new BigNumber('1124.2866831844021983137114413887526'),
      },
    };

    const action = { type: fetchBeefyBuybackAction.fulfilled, payload: payload };
    const newState = buybackSlice.reducer(initialBuybackState, action);
    expect(newState).toMatchSnapshot();

    // expect that we can compute total buyback
    const state: BeefyState = { biz: { buyback: newState } } as BeefyState;
    expect({
      totalBuybackUsdAmount: selectTotalBuybackUsdAmount(state),
      totalBuybackTokenAmount: selectTotalBuybackTokenAmount(state),
    }).toMatchSnapshot();
  });
});
