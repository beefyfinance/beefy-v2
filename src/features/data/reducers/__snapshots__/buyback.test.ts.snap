// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buyback slice tests should update state on fulfilled buyback action 1`] = `
Object {
  "byChainId": Object {
    "bsc": Object {
      "buybackTokenAmount": "0",
      "buybackUsdAmount": "0",
    },
    "cronos": Object {
      "buybackTokenAmount": "0.503019530037920513",
      "buybackUsdAmount": "1124.2866831844021983137114413887526",
    },
    "polygon": Object {
      "buybackTokenAmount": "0.535709913009709997",
      "buybackUsdAmount": "1197.3521608619992880891961859928494",
    },
  },
}
`;

exports[`Buyback slice tests should update state on fulfilled buyback action 2`] = `
Object {
  "totalBuybackTokenAmount": "1.03872944304763051",
  "totalBuybackUsdAmount": "2321.638844046401486402907627381602",
}
`;
